torch/optim/adagrad.py

```diff
@@ -50,7 +50,7 @@ def __init__(
         for group in self.param_groups:
             for p in group["params"]:
                 state = self.state[p]
-                state["step"] = torch.tensor(0.0)
+                state["step"] = torch.tensor(0.0, device=p.device)
```

Comment:
This will actually slowdown eager code, as eager code would prefer this to be in CPU. 

But it seems necessary to overcome the device mismatch in the buffers during fusion scheduling. Nevertheless, it is odd that e.g. `Adam` does not display this requirement - even if its `step` tensor is on CPU, all of its input buffers end up being on GPU.

Reply:
That was in theory - it seems that in practice there is no difference:

![image](https://github.com/pytorch/pytorch/assets/9093549/4f31def1-c5c3-48fc-b6f3-d8039bc64560)

Reply:
A comparison for `torch.compile` would also be interesting - to understand the practical speedup of having fewer kernels.